Source/missiles.cpp

```diff
@@ -2810,69 +2810,66 @@ int AddMissile(Point src, Point dst, Direction midir, missile_id mitype, mienemy
 {
-	missile._mirange--;
+	int mind;
+	int maxd;
+	int eMind;
+	int eMaxd;
 	int p = missile._misource;
+	missile_graphic_id eAnim;
+	missile_resistance eRst;
+	auto &player = Players[p];
+
+	missile._mirange--;
 	if (missile._miAnimType == MFILE_MINILTNG || missile._miAnimType == MFILE_MAGBLOS) {
 		ChangeLight(missile._mlid, missile.position.tile, missile._miAnimFrame + 5);
-		missile_resistance rst = MissilesData[missile._mitype].mResist;
-		if (missile._mitype == MIS_LARROW) {
-			int mind;
-			int maxd;
-			if (p != -1) {
-				mind = Players[p]._pILMinDam;
-				maxd = Players[p]._pILMaxDam;
-			} else {
-				mind = GenerateRnd(10) + 1 + currlevel;
-				maxd = GenerateRnd(10) + 1 + currlevel * 2;
-			}
-			MissilesData[MIS_LARROW].mResist = MISR_LIGHTNING;
-			CheckMissileCol(missile, mind, maxd, false, missile.position.tile, true);
-		}
-		if (missile._mitype == MIS_FARROW) {
-			int mind;
-			int maxd;
-			if (p != -1) {
-				mind = Players[p]._pIFMinDam;
-				maxd = Players[p]._pIFMaxDam;
-			} else {
-				mind = GenerateRnd(10) + 1 + currlevel;
-				maxd = GenerateRnd(10) + 1 + currlevel * 2;
-			}
-			MissilesData[MIS_FARROW].mResist = MISR_FIRE;
-			CheckMissileCol(missile, mind, maxd, false, missile.position.tile, true);
-		}
-		MissilesData[missile._mitype].mResist = rst;
 	} else {
 		missile._midist++;
-
-		int mind;
-		int maxd;
 		if (p != -1) {
 			if (missile._micaster == TARGET_MONSTERS) {
-				mind = Players[p]._pIMinDam;
-				maxd = Players[p]._pIMaxDam;
+				mind = player._pIMinDam;
+				maxd = player._pIMaxDam;
 			} else {
-				mind = Monsters[p].mMinDamage;
-				maxd = Monsters[p].mMaxDamage;
+				auto &monster = Monsters[p];
+				mind = monster.mMinDamage;
+				maxd = monster.mMaxDamage;
 			}
 		} else {
 			mind = GenerateRnd(10) + 1 + currlevel;
 			maxd = GenerateRnd(10) + 1 + currlevel * 2;
 		}
-
 		missile_resistance rst = MissilesData[missile._mitype].mResist;
 		MissilesData[missile._mitype].mResist = MISR_NONE;
-		MoveMissileAndCheckMissileCol(missile, mind, maxd, true, true);
+		MoveMissileAndCheckMissileCol(missile, mind, maxd, true, false);
 		MissilesData[missile._mitype].mResist = rst;
-
 		if (missile._mirange == 0) {
 			missile._mimfnum = 0;
-			if (missile._mitype == MIS_LARROW)
-				SetMissAnim(missile, MFILE_MINILTNG);
-			else
-				SetMissAnim(missile, MFILE_MAGBLOS);
 			missile._mirange = missile._miAnimLen - 1;
 			missile.position.StopMissile();
+
+			rst = MissilesData[missile._mitype].mResist;
+
+			switch (missile._mitype) {
+			case MIS_LARROW:
+				eMind = player._pILMinDam;
+				eMaxd = player._pILMaxDam;
+				eAnim = MFILE_MINILTNG;
+				eRst = MISR_LIGHTNING;
+				break;
+			case MIS_FARROW:
+				eMind = player._pIFMinDam;
+				eMaxd = player._pIFMaxDam;
+				eAnim = MFILE_MAGBLOS;
+				eRst = MISR_FIRE;
+				break;
+			}
+			SetMissAnim(missile, eAnim);
+			if (p == -1) {
+				eMind = GenerateRnd(10) + 1 + currlevel;
+				eMaxd = GenerateRnd(10) + 1 + currlevel * 2;
+			}

```

Comment:
The same point can be made here. This was a good idea to try and clean things up, but if `p == -1`, you just initialized `eMind` and `eMaxd` using an invalid player.

Reply:
Ah yes! I'll fix that tonight